[blog] Add articles about Arachne

enumag opened the pull request:
@TomasVotruba @f3l1x Can I ask you for feedback?

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+perex: '''
+    In many cases <a href="https://github.com/nette/security">Nette/Security</a> lacks the API needed for certain tasks.
+    Experienced Nette users therefore often recommend using some custom solution instead.
+    In this article I'll go over the known problems and how <a href="https://github.com/Arachne/Security">Arachne/Security</a> can help you solve them.
```
"...the known problems" - I'd extends with 2 examples, so people with the same could relate to it and open the post more likely

"...the known problems like ..., ..."

enumag replied:
I changed the headline to `Arachne/Security - Separate Authentication and Session Refresh` so since the problems are already there I don't think I need to repeat them.

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+---
+
+Separate authentication
+----
```
It's common to use `##` in here, so it would be nice to keep that standard

Also this notation confused me visually with `---` above and ````` bellow

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+Separate authentication
+----
+
+For e-shops and many other applications you will need accounts for administrators and preferably completely separate accounts for custommers. While it could be solved with authorization instead, separate authentication is often preferred.
```
Nice issue to solve! I'd highlight it with bold and bit rephrase it.

...**need completely separated accounts for administrators and for custommers**

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
````diff
@@ -0,0 +1,210 @@
+
+Nette/Security does not have a good support for this. Quite often programmers don't use the `Nette\Security\User` at all and instead work directly with `Nette\Http\UserStorage`. With Arachne/Security you can use `Arachne\Security\Authentication\Firewall` instead of `Nette\Security\User` - however unlike `Nette\Security\User` it is common to use multiple separate firewalls.
+
+```
````
You can use "yaml" here.

TomasVotruba replied:
I know what firewall is, but only thanks to Symfony. Maybe I'd describe an example with: "admin firewall, logged firewall, annonymous firewall" for Nette users. To show what it means

enumag replied:
I will add a section that Firewall is similar to Nette\Security\User with the key differences between them.

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
````diff
@@ -0,0 +1,210 @@
+{
+}
+```
+
````
I see that they were created, but don't know what to do next inside those classes.

Maybe some example?

enumag replied:
Actually nothing. These classes are for autowiring only and should stay empty in most cases.

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
````diff
@@ -0,0 +1,210 @@
+```
+
+
+Active session validation
````
In English it is comnon to use All Words With First Letter Uppercased in headlines

f3l1x commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+
+For e-shops and many other applications you will need accounts for administrators and preferably completely separate accounts for custommers. While it could be solved with authorization instead, separate authentication is often preferred.
+
+Nette/Security does not have a good support for this. Quite often programmers don't use the `Nette\Security\User` at all and instead work directly with `Nette\Http\UserStorage`. With Arachne/Security you can use `Arachne\Security\Authentication\Firewall` instead of `Nette\Security\User` - however unlike `Nette\Security\User` it is common to use multiple separate firewalls.
```
Nette\Security maybe? :-)

enumag replied:
At the start of the sentence? It refers to the package as a whole so I don't want to use backslash here.

f3l1x replied:
I see. Still, I'd prefer Nette\Security, never mind. ;-)

f3l1x commented:
It looks good to me. Good job. :+1: You deserve a :taco:

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+Active session validation
+----
+
+In Nette/Security when a user signs in his identity is valid until he signs out again or is signed out automatically because of expiration. The problem is that if you change the privileges of a user or remove his account his session will not be updated with the new privileges or terminated right away. The user will still have the same privileges he had when he signed in. Also in case an attacker  ...
```
So this is about "Refresh Session" when something changed?

enumag replied:
exactly

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+
+In Nette/Security when a user signs in his identity is valid until he signs out again or is signed out automatically because of expiration. The problem is that if you change the privileges of a user or remove his account his session will not be updated with the new privileges or terminated right away. The user will still have the same privileges he had when he signed in. Also in case an attacker  ...
+
+In Nette/Security this is considered a feature because to solve it you would have to check the database whether the session is still valid each time he sends a request. I consider this a security hole so I added an optional mechanism to Arachne/Security to deal with this.
```
I see. Nice explanation

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+
+In Nette/Security this is considered a feature because to solve it you would have to check the database whether the session is still valid each time he sends a request. I consider this a security hole so I added an optional mechanism to Arachne/Security to deal with this.
+
+The mechanism is `Arachne\Security\Authentication\IdentityValidatorInterface`. This interface gets the user's identity and should return an updated identity or null to sign him out.
```
Nice, I'd highlight the second sentence with bold, it's important.

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
````diff
@@ -0,0 +1,210 @@
+
+Here is an example how to register and implement it. It is based on Doctrine and simply signs out the user whenever his identityHash is changed. Just make sure to change the identityHash every time the user's password or privileges are changed.
+
+```
````
Here you can use yaml again

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+
+The mechanism is `Arachne\Security\Authentication\IdentityValidatorInterface`. This interface gets the user's identity and should return an updated identity or null to sign him out.
+
+Here is an example how to register and implement it. It is based on Doctrine and simply signs out the user whenever his identityHash is changed. Just make sure to change the identityHash every time the user's password or privileges are changed.
```
Nice and clear. 

I'd highlight `identityHash` as code

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+     */
+    private $repository;
+
+    public function __construct(EntityManager $em)
```
In my personal prefference, I'd use `$entityManager` to prevent supporting shortucts of 2 word naming.

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+    {
+        $entity = $this->repository->find($identity->getId());
+
+        return $entity && isset($identity->identityHash) && $identity->identityHash === $entity->getIdentityHash() ? $identity : null;
```
These is lot of `&&`, `?`, `:` and I got lost in what it does. Could you split to more functions with some naming?

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+Improved authorization
+----
+
+Authorization in Nette/Security has some long-known drawbacks as demonstrated in my 4 years old [RFC](https://forum.nette.org/cs/13458-security-iauthorizator-a-identita) (czech only). In the end my implementation was not accepted. In the comments David Grudl had a different [proposal](https://forum.nette.org/cs/13458-security-iauthorizator-a-identita#p99180) which would be far greater BC break. A ...
```
"Czech only", it's personal name

TomasVotruba replied:
Maybe mention what exact drawbacks you talk about, so I don't have to read RFC to find out.

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+
+Authorization in Nette/Security has some long-known drawbacks as demonstrated in my 4 years old [RFC](https://forum.nette.org/cs/13458-security-iauthorizator-a-identita) (czech only). In the end my implementation was not accepted. In the comments David Grudl had a different [proposal](https://forum.nette.org/cs/13458-security-iauthorizator-a-identita#p99180) which would be far greater BC break. A ...
+
+Arachne/Security introduces a new interface `Arachne\Security\Authorization\AuthorizatorInterface`. The `isAllowed` method is simplified to just resource and privilege. Implementations of this interface should use `Arachne\Security\Authentication\FirewallInterface` to get the user's identity. This way you can have multiple different authorizators for each firewall.
```
I'd use `isAllowed()` with brackets in the end to demonstrante it's a method

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
````diff
@@ -0,0 +1,210 @@
+
+Here is an example of how to register and implement the authorizator.
+
+```
````
Somewhere here I'd maybe spilt post into 2 posts. The are lot of nice features that deserves propper attention and I'm getting lost and bit tired in them right here.

You could choose more action headlines fro them with exact features they do.

```
- Arachne\Security - Security in Nette on Steroids - Firewalls and Session Refresh
- Arachne\Security - Security in Nette on Steroids - <other features>
```

enumag replied:
That could work nicely since 2 features are related to authentication and 2 to authorization. Alright I'll split it.

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
````diff
@@ -0,0 +1,210 @@
+```
+
+
+Fixed ACL
````
What exactly is fixed in ACL? I'd add it here

enumag replied:
note for me: Fixed permission callbacks in ACL

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+Fixed ACL
+----
+
+In the example above you might have noticed that I used `Arachne\Security\Authorization\Permission` instead of `Nette\Security\Permission`. This adresses the core issue from the RFC I linked above. It changes the parameters which will be recieved by callbacks passed to `allow` and `deny` methods. You can finally implement all permissions that require an identity check the right way without hacks.
```
"RFC I linked above", I'd link it here as well, so reader doesn't have to Google  it in this post

TomasVotruba replied:
Here might be useful a before/after code, to compare "Nette way" and "Arachne\Security" way.

This is common technique to teach people new methods by building a knowledge bridge from what they already know.

enumag replied:
The nette hack is so long ugly that I really don't want to post it here. Maybe I should add a link to some ancient topic on nette forum where it is. Besides the code example below is quite long already.

TomasVotruba commented on source/_posts/2017/2017-07-21-introducing-arachne-security.md:
```diff
@@ -0,0 +1,210 @@
+
+The goal of Arachne/Security is only to provide a better API then Nette/Security. It just implements what Nette/Security could not because of BC but still internally uses some of the classes and interfaces from Nette/Security.
+
+However this is actually just a small portion of how Arachne can help you implement the security layer of your application. There is another package called [Arachne/SecurityVerification](https://github.com/Arachne/SecurityVerification) which integrates Arachne/Security with [Arachne/Verifier](https://github.com/Arachne/Verifier) to greatly simplify your security layer with annotations and latte m ...
```
Nice follow up! You could actually link your next posts in here.

enumag replied:
The problem is that it would link to the next article before it is actually meant to be public (as they are separated by a week). Any suggestion about that other than making a pull request every week? (I usuallly have a followup note at the end of each article.)

TomasVotruba replied:
I'd add them in the end after the series finishes.

TomasVotruba commented on source/_posts/2017/2017-07-28-introducing-arachne-verifier.md:
```diff
@@ -0,0 +1,132 @@
+
+With [Arachne/Verifier](https://github.com/Arachne/Verifier) you can add rules to your presenters, actions, signals and components. They will be available only if all the rules are met, otherwise the request is denied.
+
+The point is that you don't need to repeat these rules in your Latte templates to determine when to show which link. There is a macro to simplify this.
```
"... you don't need to repeat these rules in your Latte templates to determine when to show which link"

I'd show a latte code example with those repeated rules. Again before/after. It makes easier to understand where I can use this in my project.

enumag replied:
@TomasVotruba Is the code regular we wrote not yet on pehapkari.cz? I added the latte example and statie refused to compile it.

TomasVotruba commented on source/_posts/2017/2017-07-28-introducing-arachne-verifier.md:
````diff
@@ -0,0 +1,132 @@
+<a n:ifLinkVerified="Article:edit $article->getId()" n:href>Edit</a>
+```
+
+If your rules follow some pattern across your application you can implement `Arachne\Verifier\RuleProviderInterface` to generate the rules automatically according to your conventions instead of repeating the rules for every action. Alternatively you can write a provider to pull rules from a neon/yaml/whatever file in case you don't like annotations.
````
It would be nice to have and example code for this. 
E.g. show second method with the same rule, then refactor it to as service.

Also, I'd put it in the end of the post, since it requires extra work from a reader.

enumag replied:
Moved to the end of the article. With code sample it should actually get a separate article so maybe later.

TomasVotruba commented on source/_posts/2017/2017-07-28-introducing-arachne-verifier.md:
```diff
@@ -0,0 +1,132 @@
+
+First install Arachne/Verifier according to [documentation](https://github.com/Arachne/Verifier/blob/master/docs/index.md).
+
+Next you'll need some rules for Verifier. Arachne has 3 packages providing some rules but you can implement your own as well.
```
I'd make **Arachne has 3 packages providing some rules** bold and remove the rest of sentence. Again moving "own rules" section to the end of the post. Here readers want to know "how to use it right away". If he or she likes it, he or she will think about extension. Not the other way around (first extension, then how to use it).

TomasVotruba commented on source/_posts/2017/2017-07-28-introducing-arachne-verifier.md:
````diff
@@ -0,0 +1,132 @@
+```
+
+
+[Arachne/ComponentsProtection](https://github.com/Arachne/ComponentsProtection)
````
I'd put a number before each package headline, just to make it clear to orientate.

### 1. package one

### 2. package two

### 3. package three

TomasVotruba commented on source/_posts/2017/2017-07-28-introducing-arachne-verifier.md:
```diff
@@ -0,0 +1,132 @@
+[Arachne/ComponentsProtection](https://github.com/Arachne/ComponentsProtection)
+----
+
+This package can be used to restrict components to certain actions to avoid [vulnerability](http://www.youtube.com/watch?v=ivDl8g0NEwg&t=57m4s) (the video is in czech) common to many Nette-based applications. Using this addon you can restrict your components to only certain actions using the `@Actions` annotation. To avoid omitting this annotation by accident it is strictly required for all compo ...
```
Could  you describe in one sentence the vulnerability?

"...common to many Nette-based applications. Basically: when happens this and this, then component ..."

enumag replied:
It is described int a comment in the example below. But ok I'll try to move it here.

TomasVotruba commented on source/_posts/2017/2017-08-04-introducing-arachne-entity-loader.md:
```diff
@@ -0,0 +1,184 @@
+The basics
+----
+
+With Arachne/EntityLoader, your presenters can look like this.
```
I'd start with "without" example. Jumping right away to the solution got me a bit confused.

By showing how painful this is to solve in native Nette application, people would love to solve their problems by using this package.

TomasVotruba commented on source/_posts/2017/2017-08-04-introducing-arachne-entity-loader.md:
````diff
@@ -0,0 +1,184 @@
+}
+```
+
+Of course you can use objects to create links in Latte. But if you have a case where you don't need the actual entities for anything and only have IDs it will work as well.
````
I'm don't understand this sentence. Could you rephrase? "But if you have a case where you don't need the actual entities for anything and only have IDs it will work as well."

TomasVotruba commented on source/_posts/2017/2017-08-04-introducing-arachne-entity-loader.md:
```diff
@@ -0,0 +1,184 @@
+The Request object
+----
+
+Unlike older solutions like [Zenify/DoctrineMethodsHydrator](https://github.com/DeprecatedPackages/DoctrineMethodsHydrator) EntityLoader will put the entities into your `Nette\Application\Request` object as well. This will make them available in more places than just your actions and signals. Like if you need to pass the product to a component.
```
Nice!

TomasVotruba commented on source/_posts/2017/2017-08-04-introducing-arachne-entity-loader.md:
````diff
@@ -0,0 +1,184 @@
+```
+
+
+The Request object
````
I'd prolong the headline with the feature so it's more catchy and explanins the added value.

E.g. **The Request Object with all the Entities** or something like that

TomasVotruba commented on source/_posts/2017/2017-08-04-introducing-arachne-entity-loader.md:
```diff
@@ -0,0 +1,184 @@
+    public function createComponentForm(): ProductForm
+    {
+        /** @var Product $product */
+        $product = $this->getRequest()->getParameter('product');
```
Cool!

TomasVotruba commented on source/_posts/2017/2017-08-04-introducing-arachne-entity-loader.md:
```diff
@@ -0,0 +1,184 @@
+
+This may not look too important. The point is that having entities in `Nette\Application\Request` also unleashes the full power of both [Arachne/ParameterValidation](https://github.com/Arachne/ParameterValidation) and [Arachne/SecurityVerification](https://github.com/Arachne/SecurityVerification).
+
+(Read my [previous article](/blog/2017/07/28/introducing-arachne-verifier) about [Arachne/Verifier](https://github.com/Arachne/Verifier) before continuing.)
```
Nice!

TomasVotruba replied:
I imagine here would be nice place for a another post, that explaing their combination. I'm not sure I'm ready to use another 2 packages combined.

enumag commented:
@TomasVotruba Thanks for your awesome review! I split and updated the articles. It's not meant to be merged as is though. I need to change the dates and add links to previous articles when we're satisfied with the text.

TomasVotruba commented:
@enumag Glad to help!
Do you commit with changes only? So I read only what changed and not whole long text again.

enumag commented:
@TomasVotruba I could do it but it would not help you too much since the articles were split from 3 to 5.

TomasVotruba commented:
@enumag It would be much easier to read diff:

- 1 commit for split (only copy-paste) 
- and 1 for changes in the content

enumag commented:
@TomasVotruba After seeing https://www.youtube.com/watch?v=JAYeMsAxZOE I asked @matej21 for feedback on this. He said he would do it this weekend so let's wait for that.

matej21 commented on source/_posts/2017/2017-07-16-introducing-arachne-entity-loader.md:
```diff
@@ -0,0 +1,177 @@
+
+class ProductPresenter extends Presenter
+{
+    public function actionDetail(Product $product): void
```
dle DOC toho balicku, to vypada, ze musim pouzit traitu, ale tady to vypada, ze se to deje nejak automaticky

enumag replied:
Traita je nutná jen pro store/restoreRequest, ale přidám ji.

matej21 commented on source/_posts/2017/2017-07-15-introducing-arachne-verifier.md:
````diff
@@ -0,0 +1,137 @@
+
+```
+{* Standard Nette way *}
+TODO remove comment
````
to TODO tu ma zustat?

enumag replied:
nn to čekalo na jiný PR, Statie mělo problém s mícháním markdown a latte

matej21 commented on source/_posts/2017/2017-07-16-introducing-arachne-entity-loader.md:
```diff
@@ -0,0 +1,177 @@
+use App\Entity\Product;
+use Nette\Application\UI\Control;
+
+class ProductFormGuide extends Control
```
tyhle ukazky kodu na me pusobi spise jako vizualni sum, nejak bych to osekal

matej21 commented on source/_posts/2017/2017-07-13-introducing-arachne-security.md:
```diff
@@ -0,0 +1,130 @@
+
+## Session Refresh
+
+In Nette/Security when a user signs in his identity is valid until he signs out again or is signed out automatically because of expiration. The problem is that if you change the privileges of a user or remove his account his session will not be updated with the new privileges or terminated right away. The user will still have the same roles he had when he signed in. Also in case an attacker steal ...
```
mozna bych presne specifikoval ten zpusob ukladani identity, neco jako "In Nette/Security when a user signs in his identity is by default serialized and stored in session and is valid until ..."

enumag commented:
@matej21 Thanks for your review!

@TomasVotruba All done, ready to merge.

TomasVotruba commented:
@enumag Sure, you've samed me some work :+1:

enumag commented:
@TomasVotruba I thought the first article would appear on the website today - but it didn't. Any idea why?

TomasVotruba commented:
@enumag Daily Travis cron job got stuck and waited 2 days, not sure why.
I restared it just now.

TomasVotruba commented:
It's there: https://pehapkari.cz/blog/2017/08/14/arachne-security-separate-authentication-and-session-refresh/

Thanks for reminder :ghost:
